File: go/matrix.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

// Matrix multiplication implementation
func matrixMultiply(matrixA, matrixB [][]float64) [][]float64 {
	rowsA := len(matrixA)
	colsA := len(matrixA[0])
	colsB := len(matrixB[0])

	// Initialize result matrix with zeros
	result := make([][]float64, rowsA)
	for i := range result {
		result[i] = make([]float64, colsB)
	}

	// Perform multiplication
	for i := 0; i < rowsA; i++ {
		for j := 0; j < colsB; j++ {
			for k := 0; k < colsA; k++ {
				result[i][j] += matrixA[i][k] * matrixB[k][j]
			}
		}
	}

	return result
}

// RunMatrixBenchmark runs matrix multiplication benchmark with 100x100 matrices
func RunMatrixBenchmark() {
	size := 100

	// Generate two random 100x100 matrices
	matrixA := make([][]float64, size)
	matrixB := make([][]float64, size)
	for i := 0; i < size; i++ {
		matrixA[i] = make([]float64, size)
		matrixB[i] = make([]float64, size)
		for j := 0; j < size; j++ {
			matrixA[i][j] = rand.Float64()
			matrixB[i][j] = rand.Float64()
		}
	}

	// Measure multiplication time
	startTime := time.Now()
	result := matrixMultiply(matrixA, matrixB)
	executionTime := time.Since(startTime).Milliseconds()

	fmt.Printf("Test: Matrix Multiplication\n")
	fmt.Printf("Matrix size: %dx%d\n", size, size)
	fmt.Printf("Execution time: %d ms\n", executionTime)
	fmt.Printf("Result sample (0,0): %.6f\n", result[0][0])
}

File: csharp/Matrix.cs
using System;
using System.Diagnostics;
using System.Linq;

namespace StackWizard
{
    /// <summary>
    /// Matrix multiplication implementation for performance benchmarking.
    /// Tests with two 100x100 matrices.
    /// </summary>
    public class Matrix
    {
        /// <summary>
        /// Generate a random matrix of given dimensions.
        /// </summary>
        public static double[,] GenerateRandomMatrix(int rows, int cols)
        {
            var random = new Random();
            var matrix = new double[rows, cols];
            for (int i = 0; i < rows; i++)
            {
                for (int j = 0; j < cols; j++)
                {
                    matrix[i, j] = random.NextDouble() * 100;
                }
            }
            return matrix;
        }

        /// <summary>
        /// Multiply two matrices using standard algorithm.
        /// </summary>
        public static double[,] Multiply(double[,] matrixA, double[,] matrixB)
        {
            int rowsA = matrixA.GetLength(0);
            int colsA = matrixA.GetLength(1);
            int colsB = matrixB.GetLength(1);

            var result = new double[rowsA, colsB];

            for (int i = 0; i < rowsA; i++)
            {
                for (int j = 0; j < colsB; j++)
                {
                    double sum = 0;
                    for (int k = 0; k < colsA; k++)
                    {
                        sum += matrixA[i, k] * matrixB[k, j];
                    }
                    result[i, j] = sum;
                }
            }

            return result;
        }

        /// <summary>
        /// Run matrix multiplication benchmark with 100x100 matrices.
        /// </summary>
        public static void RunBenchmark()
        {
            int size = 100;

            // Generate two random matrices
            var matrixA = GenerateRandomMatrix(size, size);
            var matrixB = GenerateRandomMatrix(size, size);

            // Measure multiplication time
            var watch = Stopwatch.StartNew();
            var resultMatrix = Multiply(matrixA, matrixB);
            watch.Stop();

            double executionTime = watch.Elapsed.TotalMilliseconds;

            Console.WriteLine("\n=== Matrix Multiplication Benchmark ===");
            Console.WriteLine($"Matrix size: {size}x{size}");
            Console.WriteLine($"Execution time: {executionTime:F2} ms");
            Console.WriteLine($"Result dimensions: {resultMatrix.GetLength(0)}x{resultMatrix.GetLength(1)}");
        }
    }
}
